### ISSUE TITLE
Charm building and pushing should be libs-aware

### ISSUE THREAD

**Issue by facundobatista** [canonical/charmcraft#213]
The charm's Build and Upload commands must be aware of charmlibs.

**When building**: check that all "used libraries from 3rd parties" are locally updated (with a timeout, because of potential network issues), and that the libraries are not locally modified/tainted.

**When uploading**: check that all "own libraries" are properly updated in the Store.

**Comment by facundobatista** [comment]
We could implement this as a linter.

**Comment by facundobatista** [comment]
Related: #417

**Comment by PietroPasotti** [comment]
Summarizing an internal discussion:

I feel that this behaviour should be implemented as an independent `check-libs` command.

Whether that is internally also used to raise some warnings at pack/publish time, that feels useful as well. But the output of such a check-libs command would also be very useful in CI settings. (see for example https://github.com/canonical/charming-actions/pull/47)

The command should:
- iterate through all `lib/charms` trees:
  - for each charm: compare all lib files declared in `lib/charms/<charm>/<v>/<lib>` with those published on charmhub and report any anomalies or an 'all good' status if no anomalies are found.
  - anomalies would be:
  - local lib version < remote lib version
  - remote lib version < local lib version
  - remote lib not found
 
[`jhack charm vinfo`](https://github.com/PietroPasotti/jhack#vinfo) implements a very similar logic:
![image](https://user-images.githubusercontent.com/6230162/213161565-41a2cd58-b200-480a-bfc9-245b683466d2.png)

**Comment by facundobatista** [comment]
Hello @PietroPasotti thanks for the info!

You say that *instead* of changing `pack` or `upload`, a separate command should be implemented so the developers have the information whenever they want?

A separate command is a good idea, but I think that having a linter that issues a warning when packing and any library is not updated is also good.

One detail: a fourth result would be "network / server error".

**Comment by PietroPasotti** [comment]
I'm not saying that you shouldn't get a warning on pack/upload if the aforementioned command reports that there are issues (I think you should), but I'm saying that the same logic should be also available to the end-users **in general**, not just in the specific situation when one is packing/uploading.

**Comment by PietroPasotti** [comment]
I disagree that #2252 closes this issue, as it only implements half the story.  What I was asking for above was for this to be available not only when packing but also as a standalone command. Just because it's way faster, I assume, than having to pack the charm just because I want to verify the libs are in sync.
